retry.go

```diff
@@ -25,60 +26,95 @@ func init() {
 	random = rand.New(rand.NewSource(time.Now().UnixNano()))
 }
 
+// requestGUIDKey is attached to every request against Snowflake
 const requestGUIDKey string = "request_guid"
 
-// Format of "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
-const uuidLen int = 36
+// retryCounterKey is attached to query-request from the second time
+const retryCounterKey string = "retryCounter"
+
+// requestIDKey is attached to all requests to Snowflake
+const requestIDKey string = "requestId"
 
 // This class takes in an url during construction and replace the
 // value of request_guid every time the replace() is called
 // When the url does not contain request_guid, just return the original
 // url
 type requestGUIDReplacerI interface {
 	// replace the url with new ID
-	replace() string
+	replace() *url.URL
 }
 
 // Make requestGUIDReplacer given a url string
-func makeRequestGUIDReplacer(url string) requestGUIDReplacerI {
-	startIndex := strings.Index(url, requestGUIDKey)
-	if startIndex == -1 {
-		return &transientReplacer{url}
+func makeRequestGUIDReplacer(urlPtr *url.URL) requestGUIDReplacerI {
+	_, err := url.ParseQuery(urlPtr.RawQuery)
```

Comment:
The reason to have a transientReplacer is to not even go through the url changing logic in the later retry if the url doesn't contains the requestGUID at the first place. So we might want to detect if the retryGuid key is in the query and if not just return the transientReplacer.